### PR TITLE
Sharding Pattern

### PR DESCRIPTION
**Pull request by Azureyjt**
This PR should fix issue #454.

Three types of sharding strategy are implemented:
- Range
- Lookup
- Hash

### CODE REVIEW

**Comment by Azureyjt** [comment]
@iluwatar Please kindly review this PR.

File: sharding/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+This pattern offers the following benefits:
+
+- You can scale the system out by adding further shards running on additional storage nodes.
+- A system can use off the shelf commodity hardware rather than specialized (and expensive) computers for each storage node.
+- You can reduce contention and improved performance by balancing the workload across shards.
+- In the cloud, shards can be located physically close to the users that will access the data.
+
+## Credits  
+  
+* [Cloud Design Patterns: Prescriptive Architecture Guidance for Cloud Applications - Sharding Pattern]([https://docs.microsoft.com/en-us/previous-versions/msp-n-p/dn589797(v=pandp.10)?redirectedfrom=MSDN](https://docs.microsoft.com/en-us/previous-versions/msp-n-p/dn589797(v=pandp.10)?redirectedfrom=MSDN))
```
[the remark this answers is not in the document]
Comment (Azureyjt) [reply sharding/README.md]: Fixed in 13ae8d58fa91906f53cc21a52766cecf4b948c94

File: sharding/src/main/java/com/iluwatar/sharding/App.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    manager.storeData(data1);
+    manager.storeData(data2);
+    manager.storeData(data3);
+    manager.storeData(data4);
+
+    shard1.clearData();
+    shard2.clearData();
+    shard3.clearData();
+
+    return;
```
[the remark this answers is not in the document]
Comment (Azureyjt) [reply sharding/src/main/java/com/iluwatar/sharding/App.java]: Fixed in 13ae8d58fa91906f53cc21a52766cecf4b948c94

**Comment by Azureyjt** [comment]
> Checkstyle with Google Checks reports many issues - should be resolved before merge

Hi iluwatar,

Thank you for reviewing the code. I'm sorry that I forgot to check the checkstyle report, and now all the code style issues are fixed. Please kindly check it.

**Comment by iluwatar** [comment]
@Azureyjt no problem, it's easy to miss now since Checkstyle doesn't fail the build. Once we get #1021 done, we can make it strict.
Thanks for the contribution once again, you're becoming regular here :relaxed:

**Pull request merged**